src/tomography/quantumtomography.jl

```diff
@@ -0,0 +1,293 @@
+    ρ̂ = ρ.value
+  end
+  return toitensor(ρ̂, sites)
+end
+
+
+tomography(data::Matrix{Pair{String, Int}}; method::String="linear_inversion", fillzeros::Bool=true, kwargs...) = 
+  tomography(empirical_probabilities(data; fillzeros=fillzeros), siteinds("Qubit", size(data,2)); method = method, kwargs...)
+
+tomography(data::Matrix{Pair{String, Int}}, sites::Vector{<:Index}; method::String="linear_inversion", fillzeros::Bool=true, kwargs...) = 
+  tomography(empirical_probabilities(data; fillzeros=fillzeros), sites; method = method, kwargs...)
+
+tomography(data::Matrix{Pair{String,Pair{String, Int}}}; kwargs...) = 
+  tomography(data, siteinds("Qubit", size(data,2)); kwargs...)
+
+function tomography(data::Matrix{Pair{String,Pair{String, Int}}}, sites::Vector{<:Index}; method::String="linear_inversion", fillzeros::Bool=true, kwargs...) 
+  sites_in  = addtags.(sites, "Input")
+  sites_out = addtags.(sites, "Output")
+  process_sites = Index[]
+  for j in 1:size(data,2) 
+    push!(process_sites, sites_in[j]) 
+    push!(process_sites, sites_out[j]) 
+  end
+  tomography(empirical_probabilities(data; fillzeros=fillzeros), process_sites; method = method, process = true, kwargs...)
+end
+
+
+
+
+
+
+"""
+    tomography(train_data::Matrix{Pair{String,Pair{String, Int}}}, L::LPDO;
+               optimizer::Optimizer,
+               observer! = nothing,
+               batchsize::Int64 = 100,
+               epochs::Int64 = 1000,
+               kwargs...)
+
+Run quantum process tomography using a variational model `L` to fit `train_data`.
+The model can be either a unitary circuit (MPO) or a Choi matrix (LPDO).
+
+# Arguments:
+- `train_data`: pairs of preparation/ (basis/outcome): `("X+"=>"X"=>0, "Z-"=>"Y"=>1, "Y+"=>"Z"=>0, …)`.
+ - `L`: variational model (MPO/LPDO).
+ - `optimizer`: algorithm used to update the model parameters.
+ - `observer!`: if provided, keep track of training metrics.
+ - `batch_size`: number of samples used for one gradient update.
+ - `epochs`: number of training iterations.
+ - `target`: target quantum process (if provided, compute fidelities).
+ - `test_data`: data for computing cross-validation.
+ - `outputpath`: if provided, save metrics on file.
+"""
+function tomography(
+  train_data::AbstractMatrix, L::LPDO; (observer!)=nothing, kwargs...
+)
+  # Read arguments
+  opt = get(kwargs, :optimizer, Optimisers.Descent(0.01))
+  batchsize::Int64 = get(kwargs, :batchsize, 100)
+  epochs::Int64 = get(kwargs, :epochs, 1000)
+  trace_preserving_regularizer = get(kwargs, :trace_preserving_regularizer, 0.0)
+  observe_step::Int64 = get(kwargs, :observe_step, 1)
+  test_data = get(kwargs, :test_data, nothing)
+  outputpath = get(kwargs, :fout, nothing)
+  print_metrics = get(kwargs, :print_metrics, [])
+  outputpath = get(kwargs, :outputpath, nothing)
+  outputlevel = get(kwargs, :outputlevel, 1)
+  savemodel = get(kwargs, :savemodel, false)
+  earlystop = get(kwargs, :earlystop, false)
```

Comment:
There are a lot of arguments here. Which parts are new?

Could you explain the design of `earlystop`?

Reply:
`outputlevel` is the same as ITensor. If `outputpath` is provided, it saves the observer. If `savemodel` is true, it also save the TN.
 
The early stop goes like this: if the observer is passed to `tomography` then one could add an early stop function to it (which halts the training if evaluated to be true). To avoid having to redifine this function everytime, you can pass `earlystop=true` to tomography, which adds a standard early stop criteria to the observer.

Reply:
I would kind of prefer if everything could go in the `Observer` for now, and then we add on these keyword arguments as needed. But that might be too idealistic and get too annoying.